sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -1140,9 +1167,26 @@ private class FullOuterIterator(
 
   override def advanceNext(): Boolean = {
     val r = smjScanner.advanceNext()
-    if (r) numRows += 1
+    if (r) {
+      numRows += 1
+    } else {
+      smjScanner.close()
+    }
     r
   }
 
   override def getRow: InternalRow = resultProj(joinedRow)
 }
+
+trait CloseableScanner {
+  def closeIterator(iter: RowIterator): Unit = {
+    if (iter.isInstanceOf[RowIteratorFromScala]) {
```

Comment:
I think you can use a `case` statement here to make it a little more concise, but no big deal.
Does this need to be a trait vs just a method that these various classes can call?

Reply:
Case statement's a good idea. Looks cleaner.

Do you mean a class instead of a trait? The various classes are all top level, so I'm not sure where I'd put this method instead.

Reply:
I guess I figured this is just a static method in a companion object somewhere. If there really isn't a meaningfully cleaner place, this is OK.